packages/grid/_modules_/grid/hooks/features/columns/useGridColumns.ts

```diff
@@ -43,11 +51,22 @@ export function useGridColumns(
     [props.columnTypes],
   );
 
+  /**
+   * If `initialState.columns.columnVisibilityModel` or `columnVisibilityModel` was defined during the 1st render, we are directly updating the model
+   * If not, we keep the old behavior and update the `GridColDef.hide` option (which will update the state model through the `GridColDef.hide` => `columnVisibilityModel` sync in `createColumnsState`
+   */
+  const shouldUseVisibleColumnModel = React.useRef(
+    !!props.columnVisibilityModel || !!props.initialState?.columns?.columnVisibilityModel,
+  ).current;
+
   useGridStateInit(apiRef, (state) => {
     const columnsState = createColumnsState({
       apiRef,
       columnsTypes,
       columnsToUpsert: props.columns,
+      shouldRegenColumnVisibilityModelFromColumns: !shouldUseVisibleColumnModel,
+      currentColumnVisibilityModel:
+        props.columnVisibilityModel ?? props.initialState?.columns?.columnVisibilityModel ?? {},
```

Comment:
By default, the id column in the Commodity dataset is hidden. However, in https://deploy-preview-3554--material-ui-x.netlify.app/components/data-grid/columns/#column-visibility all demos that use this dataset and have an initial state or the `columnVisibilityModel` prop, the id column is visible. With that being said, should the initial state be created also from the columns with `hide` too? In controlled mode I guess there's nothing we can do, unless include the id column in the object.

Reply:
I should hide manually the `id` column on the demo I think.
I would advise against using the `hide` field when the model is defined. The way I built it, the `hide` field is deprecated. If we use the model, it is not taken into account at all.
That way it is not breaking, but users won't be able to do dirty tricks with both the `hide` and `models`.

Reply:
Note that the brokerId column is also hidden in the dataset. This makes me think if wouldn't be better to include a default `initialState={{ columns: { columnVisibilityModel: { id: false } } }}` below. Once we drop `hide` it will break all demos.

https://github.com/mui-org/material-ui-x/blob/f581ac12a3dfc9097f3877b3334a935b5fdbfdac/packages/grid/x-data-grid-generator/src/services/tree-data-generator.ts#L109

Reply:
Demo will have to manually pass it to the `initialState` when they override it
But yes better do it now, we will forget if we don't

Reply:
Done

Reply:
Is it planned to remove the `shouldUseVisibleColumnModel` in v6?

Reply:
Yes, in v6 we would always use the model, the `hide` field would not exist anymore.